File: programmers/level2/더 맵게.py
import heapq


def solution(scoville, K):
    answer = 0
    heapq.heapify(scoville)

    while scoville[0] < K:  # 가장 작은 값이 K보다 크다면 탈출
        if len(scoville) > 1:  # 길이가 1 이상인 경우에만
            heapq.heappush(scoville, heapq.heappop(scoville) + heapq.heappop(scoville) * 2)
            answer += 1
        else:  # K이상 만드는 것 실패
            answer = -1
            break

    return answer

File: programmers/level2/더 맵게.java
import java.util.*;


class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>();

        for (int s: scoville) {
            pq.offer(s);
        }

        int num = 0;
        while (!pq.isEmpty()) {
            num = pq.poll();
            if (num < K) {
                if (pq.isEmpty()) {
                    answer = -1;
                    break;
                } else {
                    pq.offer(num + (pq.poll() * 2));
                    answer++;
                }
            } else {
                break;
            }
        }
        return answer;
    }
}